Fixes to Blob (unify way to read from BlobFile, move conf.py BLOB_SIZE  )

kaykurokawa opened the pull request:
8e70c0e
for HashBlobReader and HashBobWriter, add destructors that closes read/write handles if not closed already. Since this should not happen if the class is used properly,  log a Warn if destructor is called without handles being closed ( Hopefully we catch any missing close() calls this way)

eb40ffa
Have HashBlobReader take read handle instead of a file path as argument. If open() throws an exception, we don't need to worry about HashBlobReader being ever created ( thus no need to handle cases of HashBlobReader.close() without a read handle)   


4327f9d
906e8b0
10c4551
Removed StreamBlobDecryptor's way of reading from BlobFile by using consumer/producer pattern. Now we have StreamBlobDecryptor just utilize BlobFile.open_for_reading() to get a read handle, and use a deferToThread for decrypting from the read handle on a seperate thread. 


8f5166a
Moved the BLOB_SIZE fixed setting from conf.py, to MAX_BLOB_SIZE in blob/blob_file.py . This way prism can import lbrynet.blob classes without needing to set up the config. I think this is also just a better place for this constant as changing it pretty much breaks everything.

kaykurokawa commented:
rebased to master

kaykurokawa commented:
rebased again to master

jackrobison commented on lbrynet/cryptstream/CryptBlob.py:
```diff
@@ -67,13 +70,17 @@ def finish_decrypt():
 
-        d = self.blob.read(decrypt_bytes)
-        d.addCallback(lambda _: finish_decrypt())
+        d = threads.deferToThread(decrypt_bytes)
```
Why deferToThread?

kaykurokawa replied:
Before this PR StreamBlobDecryptor used FileSender() and the producer/consumer pattern to decrypt blobs in an asynchronous manner. 

If we dont't deferToThread, this will block the reactor till a standard read()  and write() to a normal file handle is finished ( the write() is called by write_bytes() see lbry_file/client/EncryptedFileDownloader.py EncryptedFileSaver._get_write_func() )

jackrobison replied:
I think it would be better to pass the read_handle to a FileBodyProducer and call decrypt_bytes on the result of that instead of doing the read() in the decrypt_bytes function

kaykurokawa commented on lbrynet/blob/reader.py:
```diff
@@ -1,50 +1,22 @@
-        self.write_func(data)
-        if self.streaming is False:
-            reactor.callLater(0, self.producer.resumeProducing)
-
```
This class here sets up the producer/consumer pattern

kaykurokawa commented on lbrynet/blob/blob_file.py:
```diff
@@ -142,33 +141,6 @@ def is_downloading(self):
-        if file_handle is not None:
-            d = file_sender.beginFileTransfer(file_handle, reader)
-            d.addCallback(close_self)
-        else:
```
you can see here that it uses FileSender() so that StreamBlobDecryptor can decrypt blobs asynchronously

kaykurokawa commented:
I changed it to use the FileBodyProducer instead of using deferToThread. I think that is the more twisted way to do it,  but it will block on the write, which probably isn't that big of a deal. 

I think the next step is to write the consumer, which will prevent blocking on the write. This will require some refactoring of the StreamBlobDecryptor. 

Rebased to master.

lyoshenka commented:
@kaykurokawa is this done?

jackrobison commented:
LGTM

kaykurokawa commented:
rebased to master and merging
